flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/DefaultExecutionGraph.java

```diff
@@ -1279,10 +1279,14 @@ private void releasePartitions(final List<IntermediateResultPartitionID> releasa
 
             // Remove cached ShuffleDescriptor when partition is released
             releasablePartitions.stream()
-                    .map(IntermediateResultPartitionID::getIntermediateDataSetID)
+                    .map(edgeManager::getConsumedPartitionGroupsById)
+                    .flatMap(List::stream)
                     .distinct()
-                    .map(intermediateResults::get)
-                    .forEach(IntermediateResult::notifyPartitionChanged);
+                    .forEach(
+                            group ->
+                                    intermediateResults
+                                            .get(group.getFirst().getIntermediateDataSetID())
+                                            .notifyPartitionChanged(group));
```

Comment:
Why is it possible that we release all consumer groups based on a released partition? If I understand the code correctly, then it should be possible that a single result partition belongs to one or more `ConsumerPartitionGroups`. If this is correct, then all consuming `ConsumerVertexGroups` need to finish in order to be allowed to remove the cached entries for every `ConsumerPartitionGroups`. Is this guaranteed at this point?

Reply:
> Why is it possible that we release all consumer groups based on a released partition? If I understand the code correctly, then it should be possible that a single result partition belongs to one or more ConsumerPartitionGroups. If this is correct, then all consuming ConsumerVertexGroups need to finish in order to be allowed to remove the cached entries for every ConsumerPartitionGroups. Is this guaranteed at this point?

> I think I don't fully understand the contracts of the releasePartitions method. From a quick glance, it does not look correct that we release the cached ShuffleDescriptors of all ConsumerPartitionGroups based on a released IntermediateResultPartitionID unless we can guarantee that releasablePartitions only contains partitions that belong to ConsumerPartitionGroups that are fully consumed.

I think it's guaranteed semantically. 

A result partition can be released only when all its consumers are finished. A ConsumedPartitionGroup contains all the isomorphic partitions that be consumed by the same consumer vertices. And these consumer vertices forms a ConsumerVertexGroup (illustrated in FLINK-21110).

Therefore, if a partition is releasable, all consuming ConsumerVertexGroups are surely finished. That means the ConsumedPartitionGroups corresponding to these ConsumerVertexGroups are releasable. Thus, We can remove the cache for them.

Besides, once a partition is released, it doesn't exist any longer. Thus the cache for the ConsumedPartitionGroups it belongs are not valid anymore. They should be removed.

I'll add more comments in the code here.

The semantic of "If a partition is releasable, all consuming ConsumerVertexGroups are surely finished, the corresponding ConsumedPartitionGroups are releasable" is guaranteed by `PartitionReleaseStrategy#vertexFinished`. Maybe we should change the interface `List<IntermediateResultPartitionID> vertexFinished(ExecutionVertexID finishedVertex)` into `List<ConsumedPartitionGroup> vertexFinished(ExecutionVertexID finishedVertex)`? In this way, we can make sure the whole ConsumedPartitionGroup is released together.

> I think part of the problem for me is the constant mapping between partitions, consumer partition groups and consumer vertex groups at various places in the code. I have the feeling that this logic is a bit too spread to be easily understandable.

Yes, I agree that the concept of ConsumedPartitionGroup and ConsumerVertexGroup is not easily understandable. But in order to decrease the computation complexity of traversing the graph and optimize the performance of scheduler, I think it's necessary to group the isomorphic vertices and partitions. I'm wondering, would it be better if I write a detailed document to explain the concept of ConsumedPartitionGroup and ConsumerVertexGroup?

Reply:
Obviously, part of the problem is me not being fully immersed into the details of this code. I do understand that we need extended bookkeeping logic to execute things more efficiently and the groups are a good solution for it. 

I am mainly wondering whether it would be possible to structure the code a bit differently so that related concepts are located a bit closer together. Maybe my point becomes a bit clearer when looking at the control flow:

1. An `Execution` terminates, this triggers `maybeReleasePartitions`
2. `maybeReleasePartitions` checks via `partitionReleaseStrategy.vertexFinished` which partitions are releaseable
3. Internally, partitionReleaseStrategy looks up which `ConsumerPartitionGroups` are fully consumed.
4. Based on the `ConsumerPartitionGroups` we return which `IntermediateResultPartitionID` can be released and call `releasePartitions`
5. Release partitions tries to rediscover the `ConsumerPartitionGroups` based on the `IntermediateResultPartitionID` whose `ShuffleDescriptors` can be released.
6. We `stopTrackingAndReleasePartitions`

In particular, step 5. seems overly complicated because of the reverse lookup between `IntermediateResultPartitionID` and `ConsumerPartitionGroup`. I think we already have this information in step 3./4.

Part of the problem is probably that we added the `EdgeManager` and the group logic after everything else was added. Maybe it would help to change `PartitionReleaseStrategy` to return the set of completed `PartitionConsumerGroups`. Then one could clear that cached `ShuffleDescriptors` and then call `releasePartitions` with the partitions contained in the `PartitionConsumerGroups`.

Reply:
Thank you, @tillrohrmann, for this detailed explanation! I think my next step is to add a hotfix that changes `PartitionReleaseStrategy` to return the set of completed `ConsumedPartitionGroups`. 

Furthermore, a hotfix to get rid of the assumption that one IntermediateDataSet can have multiple consumer `JobVertices` will also be added, as we mentioned [above](https://github.com/apache/flink/pull/16856#discussion_r692056461).

Reply:
It is just an idea. Maybe the code can be changed also in a different way to make the control flow a bit clearer. So if you see anything then feel free to go for it.